eng/tools/mgmtreport/main.go

```diff
@@ -0,0 +1,391 @@
+	flag.StringVar(&organizationUrl, "organization", "https://dev.azure.com/azure-sdk", "Azure Devops Organization Url")
+	flag.StringVar(&projectName, "project", "internal", "Azure Devops Project Name")
+}
+
+func main() {
+	flag.Parse()
+
+	if sdkPath == "" {
+		flag.PrintDefaults()
+		log.Fatal("Please enter the azure-sdk-for-go path")
+	}
+	if storageAccountName == "" {
+		flag.PrintDefaults()
+		log.Fatal("Please enter the Azure Storage Account Name")
+	}
+	storageAccountKey, ok := os.LookupEnv("AZURE_STORAGE_PRIMARY_ACCOUNT_KEY")
+	if !ok {
+		log.Fatal("AZURE_STORAGE_PRIMARY_ACCOUNT_KEY could not be found")
+	}
+
+	personalAccessToken, ok := os.LookupEnv("AZURE_DEVOPS_PERSONAL_ACCESS_TOKEN")
+	if !ok {
+		log.Fatal("AZURE_DEVOPS_PERSONAL_ACCESS_TOKEN could not be found")
+	}
+
+	log.Printf("start running in %s...\n", sdkPath)
+	startTime := time.Now()
+	mgmtReport, err := execute(sdkPath, personalAccessToken)
+	if err != nil {
+		log.Fatal(err)
+	}
+	log.Println("Statistical mgmt report time:", time.Since(startTime))
+
+	log.Println("Write the mgmt report to the mgmtreport.md file...")
+	htmlData, err := writeMgmtFile(mgmtReport, path.Join(sdkPath, "mgmtReport.md"))
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if htmlData != nil {
+		log.Println("Upload mgmt report to storage account...")
+		err = uploadMgmtReport(htmlData, storageAccountName, storageAccountKey, storageContainerName, containerBlobName)
+		if err != nil {
+			log.Fatal(err)
+		}
+	}
+}
+
+func execute(sdkPath, personalAccessToken string) (map[string]mgmtInfo, error) {
+	conn := azuredevops.NewPatConnection(organizationUrl, personalAccessToken)
+	testClient, err := test.NewClient(ctx, conn)
+	if err != nil {
+		return nil, err
+	}
+
+	buildClient, err := build.NewClient(ctx, conn)
+	if err != nil {
+		return nil, err
+	}
+
+	azureDevopsClient := azuredevops.NewClient(conn, organizationUrl)
+
+	pipelineClient := pipelines.NewClient(ctx, conn)
+	pipelinesList, err := pipelineClient.ListPipelines(ctx, pipelines.ListPipelinesArgs{
+		Project: &projectName,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// filter pipelineList
+	pipelinesMap := make(map[string]*pipelines.Pipeline)
+	for i, pipe := range *pipelinesList {
+		if *pipe.Folder == "\\go" && !strings.Contains(*pipe.Name, "weekly") && strings.Contains(*pipe.Name, "go - arm") {
+			pipelinesMap[*pipe.Name] = &(*pipelinesList)[i]
+		}
+	}
+
+	// read all module
+	sdkPath = strings.ReplaceAll(sdkPath, "\\", "/")
+	modulePath := filepath.Join(sdkPath, "/sdk/resourcemanager")
+	dirs, err := os.ReadDir(modulePath)
+	if err != nil {
+		return nil, err
+	}
+
+	mgmtReport := make(map[string]mgmtInfo)
+	for _, dir := range dirs {
+		if dir.IsDir() && dir.Name() != "internal" {
+			armDirs, err := os.ReadDir(filepath.Join(modulePath, dir.Name()))
+			if err != nil {
+				return nil, err
+			}
+
+			for _, arm := range armDirs {
+				moduleName := fmt.Sprintf("%s/%s", dir.Name(), arm.Name())
```

Comment:
Could you please split all the step into several functions that could make the `execute` function shorter and clearer?